### PR TITLE
btl/base: add subsystem to support for active-message based RDMA/atomic emulation

### PR DESCRIPTION
**Pull request by hjelmn**

### CODE REVIEW

**Comment by hjelmn** [comment]
Needed to provided working RMA when no RDMA-capable device is present.

**Comment by awlauria** [comment]
bot:aws:retest

**Comment by awlauria** [comment]
Is this series of commits intended to fix https://github.com/open-mpi/ompi/issues/7830 ?

File: ompi/mca/osc/rdma/osc_rdma.h [hunk 1]
```diff
@@ -55,6 +55,8 @@
 
 #define RANK_ARRAY_COUNT(module) ((ompi_comm_size ((module)->comm) + (module)->node_count - 1) / (module)->node_count)
 
+#define MCA_OSC_RDMA_MAX_BTLS 2
```
Comment (devreal) [inline ompi/mca/osc/rdma/osc_rdma.h:58]: Maybe to make this constant more self-describing:
```suggestion
#define MCA_OSC_RDMA_MAX_USED_BTLS 2
```

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 2]
```diff
@@ -0,0 +1,1064 @@
...
+    size_t seg0_len = segments[0].seg_len - skip_bytes;
+
+    if (seg0_len > 0) {
+        BTL_VERBOSE(("unpacking %" PRIsize_t " bytes from segment 0 to %p", seg0_len, (void *) addr));
+        memcpy ((void *) addr, seg0_data, seg0_len);
+        addr += seg0_len;
+    }
+
+    for (size_t i = 1 ; i < segment_count ; ++i) {
+	size_t seg_len = segments[i].seg_len;
```
Comment (devreal) [inline opal/mca/btl/base/btl_base_am_rdma.c:300]: Indentation in this block is off

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 3]
```diff
@@ -0,0 +1,1064 @@
...
+        BTL_VERBOSE(("packing %" PRIsize_t " bytes from 0x%" PRIx64 " to segment 0",
+                     seg0_len, addr));
+        memcpy (seg0_data, (const void *) addr, seg0_len);
+        addr += seg0_len;
+        max_len -= seg0_len;
+        segments[0].seg_len = seg0_len + skip_bytes;
+    }
+
+    for (size_t i = 1 ; i < segment_count && max_len ; ++i) {
+	size_t seg_len = size_t_min (segments[i].seg_len, max_len);
```
Comment (devreal) [inline opal/mca/btl/base/btl_base_am_rdma.c:332]: Indentation

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 4]
```diff
@@ -0,0 +1,1064 @@
...
+
+static void mca_btl_base_rdma_queue_operation (mca_btl_base_module_t *btl, struct mca_btl_base_endpoint_t *endpoint,
+                                               mca_btl_base_descriptor_t *descriptor, uint64_t atomic_response,
+                                               const mca_btl_base_rdma_hdr_t *hdr, mca_btl_base_rdma_operation_t *operation) {
+    if (NULL == operation) {
+        operation = mca_btl_base_rdma_alloc_operation (btl, endpoint, descriptor, hdr);
+        if (NULL == operation) {
+            /* couldn't even allocate a small amount of memory. not much else can be done. */
+            BTL_ERROR(("could not allocate memory to queue active-message RDMA operation"));
+            abort();
```
Comment (devreal) [inline opal/mca/btl/base/btl_base_am_rdma.c:557]: Is there a more graceful way then to call abort here?

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 5]
```diff
@@ -0,0 +1,1064 @@
...
+}
+
+static void mca_btl_base_am_process_rdma (mca_btl_base_module_t *btl, const mca_btl_base_receive_descriptor_t *desc)
+{
+    /* not all btls work with these active message atomics. at this time
+     * all of the affected btls already have atomic support so there is
+     * no need to attempt to get the endpoint here. */
+    if (NULL == desc->endpoint) {
+        BTL_ERROR(("BTL is not compatible with active-message RDMA"));
+        abort();
```
Comment (devreal) [inline opal/mca/btl/base/btl_base_am_rdma.c:862]: Can we be more graceful here?

File: ompi/mca/osc/rdma/osc_rdma.h [hunk 6]
```diff
@@ -251,8 +256,12 @@ struct ompi_osc_rdma_module_t {
     opal_mutex_t peer_lock;
 
 
-    /** BTL in use */
-    struct mca_btl_base_module_t *selected_btl;
+    /** BTL(s) in use */
+    struct mca_btl_base_module_t *selected_btls[MCA_OSC_RDMA_MAX_BTLS];
```
Comment (devreal) [inline ompi/mca/osc/rdma/osc_rdma.h:260]: Can you add a comment which btl is used for what? It looks like memory is registered with `selected_btls[0]`  but what is the second btl for?

**Review by devreal — CHANGES_REQUESTED** [review]
Looks good to me, except for a few nit-picks.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (jsquyres) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: Yes, please do not call `abort()` -- that can lead to terrabytes of corefiles.  ☹️

File: ompi/mca/osc/rdma/osc_rdma.h (hunk 6 above)
Comment (hjelmn) [reply ompi/mca/osc/rdma/osc_rdma.h:260]: Sure. Will add a comment. Memory registration is only supported if one BTL is in use. This helps simplify the logic.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: It is either abort or SEGV. IMHO abort is much more graceful. If we get here things are bad and we will abort or SEGV somewhere if we can't even allocate 64 bytes.

Comment (jsquyres) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: `abort()` drops a corefile -- that's kinda anti-social.

If you need to exit, please call the btl error function.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 3 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:332]: looks like tabs from before indent-tabs-mode:nil took effect. Fixed.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 2 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:300]: looks like tabs from before indent-tabs-mode:nil took effect. Fixed.

File: ompi/mca/osc/rdma/osc_rdma.h (hunk 6 above)
Comment (hjelmn) [reply ompi/mca/osc/rdma/osc_rdma.h:263]: Done.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: Already called. In that case abort is not even reached.

Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: Wait. BTL_ERROR doesn't exit so we still want it to exit here. It will SEGV which will dump core files.

File: ompi/mca/osc/rdma/osc_rdma.h (hunk 1 above)
Comment (hjelmn) [reply ompi/mca/osc/rdma/osc_rdma.h]: Done.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: So, I can remove the aborts() but the semantics of this error will not change. It will print the error message then crash when operation is dereferenced.

**Comment by awlauria** [comment]
I resolved the conflict in the copyright. Let's move to get this in/verified.

**Review by devreal — APPROVED** [review]
Thanks for the fixes, LGTM

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (bwbarrett) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: I'm not sure I agree with @jsquyres on the abort thing; I think this is precisely when we want a core file (and, I believe, we call abort elsewhere).  This is a should "never, ever happen" type thing, not a "we got our configuration wrong" type thing.

**Review by bwbarrett — CHANGES_REQUESTED** [review]
we should rebase the branch and fix up the conflicts, not create another merge commit.  @awlauria please undo this.

**Comment by awlauria** [comment]
@bwbarrett I agree. However this pr has been open for weeks, I did it for expediency. 

If we want this rebased, please ping @hjelmn to do so in a timely manner.

File: opal/mca/btl/sm/btl_sm_get.c [hunk 7]
```diff
@@ -48,6 +48,9 @@ int mca_btl_sm_get_xpmem (mca_btl_base_module_t *btl, mca_btl_base_endpoint_t *e
     mca_rcache_base_registration_t *reg;
     void *rem_ptr;
 
+    BTL_VERBOSE(("mca_btl_sm_get_xpmem called with remote address 0x%" PRIx64 ", size %" PRIsize_t,
```
Comment (bwbarrett) [inline opal/mca/btl/sm/btl_sm_get.c:51]: unrelated change?

File: opal/mca/btl/tcp/btl_tcp_endpoint.c [hunk 8]
```diff
@@ -1101,7 +1101,9 @@ static void mca_btl_tcp_endpoint_send_handler(int sd, short flags, void* user)
             /* if required - update request status and release fragment */
             OPAL_THREAD_UNLOCK(&btl_endpoint->endpoint_send_lock);
             assert( frag->base.des_flags & MCA_BTL_DES_SEND_ALWAYS_CALLBACK );
-            frag->base.des_cbfunc(&frag->btl->super, frag->endpoint, &frag->base, frag->rc);
+            if (NULL != frag->base.des_cbfunc) {
```
Comment (bwbarrett) [inline opal/mca/btl/tcp/btl_tcp_endpoint.c:1104]: should this be in a later PR or its own PR?

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 9]
```diff
@@ -0,0 +1,1064 @@
...
+    queued_descriptor->btl = btl;
+    queued_descriptor->endpoint = endpoint;
+    queued_descriptor->descriptor = descriptor;
+
+    OPAL_THREAD_SCOPED_LOCK(&default_module.mutex,
+                            opal_list_append (&default_module.queued_initiator_descriptors,
+                                              &queued_descriptor->super));
+}
+
+static inline
```
Comment (bwbarrett) [inline opal/mca/btl/base/btl_base_am_rdma.c:361]: your function declaration format inconsistencies make my OCD hurt...

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 10]
```diff
@@ -0,0 +1,1064 @@
...
+
+    return OPAL_SUCCESS;
+}
+
+static void mca_btl_base_am_descriptor_complete (mca_btl_base_module_t *btl,
+                                                 struct mca_btl_base_endpoint_t *endpoint,
+                                                 mca_btl_base_descriptor_t *descriptor, int status)
+{
+    mca_btl_base_rdma_context_t *context = (mca_btl_base_rdma_context_t *) descriptor->des_context;
+
```
Comment (bwbarrett) [inline opal/mca/btl/base/btl_base_am_rdma.c:415]: extra newline?

**Review by bwbarrett — APPROVED** [review]
approving, although the header issue makes me really nervous.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (jsquyres) [reply opal/mca/btl/base/btl_base_am_rdma.c:557]: Maybe we should have an option for dropping corefiles.  It is very anti-social to have an MTT run fill up filesystems with terrabytes of corefiles.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 9 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c]: C&P from multiple sources with different formatting. Get too used to ```p4 fix```. I created a clang-format style file and ran the new files through it. Should be completely consistent now. Will open a PR for adding the clang-format file to contrib.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 10 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Done.

File: opal/mca/btl/sm/btl_sm_get.c (hunk 7 above)
Comment (hjelmn) [reply opal/mca/btl/sm/btl_sm_get.c]: Debug code that made it into the PR. Fixed.

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 11]
```diff
@@ -0,0 +1,1064 @@
...
+    /** rdma/atomic operation is queued for response */
+    bool is_queued;
+    /** rdma operation was completed (waiting response) */
+    bool is_completed;
+};
+typedef struct mca_btl_base_rdma_operation_t mca_btl_base_rdma_operation_t;
+
+static OBJ_CLASS_INSTANCE(mca_btl_base_rdma_operation_t, opal_list_item_t, NULL, NULL);
+
+static inline size_t size_t_min (size_t a, size_t b) {
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Agreed. Can do that in a different PR.

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 12]
```diff
@@ -0,0 +1,1064 @@
...
+static OBJ_CLASS_INSTANCE(mca_btl_base_am_rdma_queued_descriptor_t, opal_list_item_t, NULL, NULL);
+
+/**
+ * @brief header for an active-message atomic/RDMA operation
+ */
+struct mca_btl_base_rdma_hdr_t {
+    /** type of operation requested. */
+    uint8_t type;
+
+    union {
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c:145]: Fixed. Added padding where necessary.

**Review by bwbarrett — APPROVED** [review]

File: opal/mca/btl/tcp/btl_tcp_endpoint.c (hunk 8 above)
Comment (hjelmn) [reply opal/mca/btl/tcp/btl_tcp_endpoint.c:1104]: Moved it into its own commit. Needed for this PR because the tcp btl force sets this flag ignoring the intent of the caller. It is reasonable to expect the caller doesn't care when the frag was actually sent so a NULL callback is reasonable.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 4 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Shouldn't happen ever in MTT until you are expecting to completely run out of memory. But, again, in that case we really can't control whether you core or not. Up to your shell to set ```ulimit -c 0```.

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 5 above)
Comment (hjelmn) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Resolving as this is the same as the discussion above.

**Comment by awlauria** [comment]
bot:aws:retest

**Comment by gpaulsen** [comment]
ubuntu executor died with a java exception.

**Comment by gpaulsen** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
I still see 6 uses of `abort()` on this PR, and I don't know if I agree with all of them.  It looks like some of the uses of `abort()` in this PR are really assert-like checks -- i.e., "if we got here, that's an OMPI programmer error". Other `abort()`s are set for run-time errors outside the scope of Open MPI, such as network packet corruption.  But a corefile on just the receiver (for example) without a corresponding corefile on the sender -- where the assert-like-check won't fail -- may not reveal anything useful.  They'll just take up space on the filesystem.

Outside the scope of this PR, I wonder if we should make `opal_assert()` that would behave exactly like `assert()`, but it has the _option_ to drop a corefile or not (e.g., via MCA param or somesuch).

I'm really sensitive to this whole corefile issue because there have been many times over the years where MTT caused the shared filesystem I use for Cisco MTT to fill up with terrabytes/petabytes/whateverbytes of corefiles.  My colleagues get (rightfully) pissed at me when this happens.

It could be helpful if one could optionally turn on corefiles from `opal_assert()` because this would let me leave `ulimit -c unlimited` to catch the *random* failures where a corefile can definitely be helpful, but are far harder to isolate and reproduce.  If we get large-scale `opal_assert()` failures and a corefile is needed for further analysis, it would seem statistically probable that running again with core dumps enabled would yield what is needed.

**Comment by bwbarrett** [comment]
Nathan's right; there's not a great way for OMPI to decide whether to drop a core file or not.  That's really a system setting (via ulimit).  We use abort() or assert() or similar so liberally in the OMPI code base, that this seems like a really weird place to hold the line.

**Comment by rhc54** [comment]
Given that Cisco isn't running MTT and the folks who are running MTT approve this PR, let's separate the question of what to do with abort (dealing with it in #8557) and get this in so MTT quits failing for those who are running it.

**Comment by jsquyres** [comment]
> Given that Cisco isn't running MTT and the folks who are running MTT approve this PR, let's separate the question of what to do with abort (dealing with it in #8557) and get this in so MTT quits failing for those who are running it.

Point of fact: Cisco *is* running MTT, but it's currently in "trial" mode (i.e., results are not visible by default) because I had some issues with networking and equipment being powered off.  Also, I objected to the use of `abort()` on this PR for two reasons, not just the "filling up of filesystems" reason.  I'll reply more on #8557.
